### PR TITLE
[Model] Add Voxtream2 1-stage TTS support

### LINKED ISSUE #1962: [New Model]: VoXtream2 (herimor/voxtream2)
### The model to consider.

https://huggingface.co/herimor/voxtream2

**VoXtream2** — a 0.5B parameter zero-shot full-stream Text-to-Speech model with dynamic speaking-rate control.

- **Architecture**: Fully autoregressive, codec-based TTS with three transformer components:
  - Incremental Phoneme Transformer (streaming phoneme processing)
  - Temporal Transformer (semantic + duration tokens via monotonic alignment)
  - Depth Transformer (acoustic codec token generation)
- **Audio codec**: Mimi (Kyutai), 16 codebooks
- **Speaker encoder**: ReDimNet for zero-shot voice cloning from 3-10s audio prompt
- **Output**: 24kHz streaming WAV audio
- **VRAM**: 2.2-4.2 GB
- **Latency**: 74ms first-packet, 4x faster than real-time on RTX 3090
- **Key features**: Dynamic mid-utterance speaking-rate control via distribution matching + classifier-free guidance, textless voice prompting (prompt-text masking)
- **Paper**: [arXiv:2603.13518](https://arxiv.org/abs/2603.13518) (Interspeech 2026 submission)
- **License**: CC-BY-4.0 (weights), MIT (code)

### The closest model vllm-omni already supports.

**Qwen3-TTS** — both are autoregressive codec-based streaming TTS models. Key differences:
- VoXtream2 uses Mimi codec (16 codebooks) vs Qwen3-TTS's custom codec
- VoXtream2 has a 3-stage transformer pipeline (phoneme → temporal → depth) vs Qwen3-TTS's talker + code predictor
- VoXtream2 is significantly smaller (0.5B vs 1.7B) with lower VRAM requirements
- VoXtream2 supports dynamic speaking-rate control mid-utterance

### What's your difficulty of supporting the model you want?

- **Multi-stage pipeline**: The 3-transformer architecture (phoneme → temporal → depth) maps naturally to vllm-omni's multi-stage pipeline, similar to how Qwen3-TTS uses Stage 0 (Talker) and Stage 1 (Code2Wav).
- **Custom codec**: Uses Mimi codec with 16 codebooks — would need a decoder stage similar to Qwen3-TTS's Code2Wav stage.
- **Speaker encoder**: ReDimNet speaker encoder for zero-shot cloning needs to run as a preprocessing step on the reference audio.
- **Phonemizer dependency**: Requires eSpeak NG for phoneme conversion as a preprocessing step.
- **Streaming**: The model is designed for full-stream inference (74ms TTFP), which aligns well with vllm-omni's streaming audio API (`/v1/audio/speech` with `stream=True`).
- **CFG (classifier-free guidance)**: Uses CFG across conditioning signals — may need special handling during batched inference.

### Use case and motivation

- **Ultra-low latency TTS**: 74ms first-packet latency makes it ideal for real-time conversational AI and voice assistants.
- **Extremely lightweight**: Only 2.2-4.2 GB VRAM — can run on consumer GPUs and be co-located with LLMs on the same device.
- **Zero-shot voice cloning**: Clone any voice from a short (3-10s) audio prompt without fine-tuning, with no transcript needed for the prompt audio.
- **Dynamic speaking rate**: Unique ability to adjust speaking speed mid-utterance — useful for expressive speech synthesis and accessibility.
- **Open weights + open code**: CC-BY-4.0 weights + MIT code, fully reproducible with published training data.

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ccczzc**
## Purpose

  Add initial Voxtream2 support in vLLM-Omni. Closes [#1962 ](https://github.com/vllm-project/vllm-omni/issues/1962)

  This PR adds:

  - Voxtream2 HF config registration and model registry entry.
  - A 1-stage Omni stage config for Voxtream2 TTS.
  - Dynamic Voxtream package/config resolution from `VOXTREAM2_ROOT`, `VOXTREAM_ROOT`, `VLLM_OMNI_VOXTREAM_CODE_PATH`, local `./voxtream`, or an installed `voxtream` package.
  - OpenAI-compatible `/v1/audio/speech` serving support with `ref_audio` handling for HTTP URLs, base64 data URLs, and scoped local `file://` URIs.
  - Offline inference and online serving examples.
  - Offline/online e2e coverage for Voxtream2.

  ### Current Limitations

  This PR intentionally supports Voxtream2 as a 1-stage end-to-end TTS path only. It wraps the upstream `voxtream.SpeechGenerator` generation loop and returns final audio from one Omni stage.

  A full native vLLM implementation is left for follow-up because the Voxtream2 runtime is not a standard single `LlamaForCausalLM` path. The upstream implementation builds a custom `Model` from three torchtune Llama 3.2 transformer decoders:

  - `phone_former`: consumes phoneme tokens and produces phoneme embeddings with dynamic look-ahead masks.
  - `temp_former`: autoregressively predicts the semantic/codebook-0 token while using a custom sink-attention context over audio history.
  - `dep_former`: predicts the remaining acoustic codebooks inside each frame and resets its KV cache every generated frame.

  These components are customized before use: token embeddings and output heads are replaced with `Identity`, and Voxtream adds its own phone/audio embeddings, speaker projection, semantic head, acoustic heads, audio-codebook shifts, dynamic masks, sink-attention state, and sampling logic around them. Because of that, the three transformers are not currently drop-in HF/vLLM CausalLM modules.

  The generation loop also couples several nontrivial runtime states that need explicit Omni stage contracts before a multi-stage native implementation:

  - Reference audio is encoded through Mimi and a separate speaker encoder, with optional prompt enhancement/VAD, delayed audio tokens, and CFG batch expansion.
  - Text is normalized/phonemized, mapped into phoneme tokens, and repeatedly converted into frame-level phone embedding indices.
  - Each generated frame first samples the semantic token and speaking-rate shift from `temp_former`, then loops through `dep_former` to sample the remaining codebooks.
  - Runtime state is updated frame by frame via `phoneme_index_map`, repeat counters, EOS index, `mimi_codes`, `sem_code`, and Mimi streaming decode.
  - Speaking-rate control changes the semantic token distribution through a rolling frame window.
  - Voxtream patches torchtune KV cache behavior for CUDA graph capture and uses custom cache reset behavior for `temp_former`/`dep_former`.

  Making this fully native would require more than registering three model classes. We need to define stable intermediate contracts for phoneme embeddings, semantic token/state, acoustic codebooks, Mimi decode state, CFG batch semantics, cache ownership, and frame-to-frame scheduler control. It would also require scheduler/orchestrator support for partial codebook/audio streaming and backpressure between stages.

  The follow-up plan is to first factor the current `SpeechGenerator` path into explicit stage boundaries, then register the three customized Llama 3.2 components as vLLM-compatible model stages, and finally add chunk/streaming support with dedicated correctness and performance tests.

  ## Test Plan
  Test on: NVIDIA GeForce RTX 4090, Driver Version: 535.288.01,  CUDA Version: 12.2 
  Unit and syntax checks:

  ```bash
  python -m py_compile \
    vllm_omni/model_executor/models/voxtream2/voxtream2.py \
    vllm_omni/entrypoints/openai/serving_speech.py \
    examples/offline_inference/voxtream2/end2end.py \
    examples/online_serving/voxtream2/speech_client.py \
    tests/e2e/offline_inference/test_voxtream2.py \
    tests/e2e/online_serving/test_voxtream2.py \
    tests/entrypoints/openai_api/test_serving_speech.py

  pytest -q tests/entrypoints/openai_api/test_serving_speech.py \
    -k "voxtream2 or prepare_speech_generation"

  pytest -q tests/e2e/online_serving/test_voxtream2.py --collect-only

  bash -n examples/online_serving/voxtream2/run_server.sh
  ```

  Manual offline benchmark:
  ```bash
  VLLM_WORKER_MULTIPROC_METHOD=spawn VOXTREAM2_ROOT=voxtream \
  python tests/e2e/offline_inference/test_voxtream2.py \
    --model herimor/voxtream2 \
    --stage-configs-path vllm_omni/model_executor/stage_configs/voxtream2_1stage.yaml \
    --voxtream-root voxtream
  ```

  Manual online serving:
  ```bash
  VLLM_WORKER_MULTIPROC_METHOD=spawn VOXTREAM2_ROOT=voxtream \
  vllm serve herimor/voxtream2 \
    --stage-configs-path vllm_omni/model_executor/stage_configs/voxtream2_1stage.yaml \
    --omni \
    --port 8091 \
    --trust-remote-code \
    --enforce-eager

  curl -X POST http://localhost:8091/v1/audio/speech \
    -H "Content-Type: application/json" \
    -d '{
      "input": "This is a Voxtream2 online serving example.",
      "ref_audio": "file:///home/zechaochen/Develop/vllm-omni/voxtream/assets/audio/english_female.wav"
    }' --output voxtream2.wav
```

  ## Test Result

  Unit and collection checks passed:

  pytest -q tests/entrypoints/openai_api/test_serving_speech.py -k "voxtream2 or prepare_speech_generation"
  10 passed, 117 deselected

  pytest -q tests/e2e/online_serving/test_voxtream2.py --collect-only
  1 test collected

  Online serving was manually verified with /v1/audio/speech and produced voxtream2.wav successfully.

  Offline benchmark result:

  completed: 10 / 10
  failed: 0
  mean RTF: 0.629
  median RTF: 0.482
  p99 RTF: 1.203
  audio throughput: 1.856x realtime

  Documentation/examples were updated for offline inference and online serving. mkdocs serve was not run locally.

### CODE REVIEW

**Review by lishunyang12 — DISMISSED** [review]
## Review: [Model] Add Voxtream2 1-stage TTS support

Good overall structure -- the 1-stage wrapper approach is well-motivated and the PR description is thorough. However, there are several issues ranging from critical (online serving path is broken) to moderate (unintended regressions on VoxCPM2, dead code, global side effects).

---

### Critical: Voxtream2 online serving path is broken -- `_detect_tts_model_type` never returns `"voxtream2"`

`_detect_tts_model_type()` in `serving_speech.py` has no branch for `_VOXTREAM2_TTS_MODEL_STAGES`. It will return `None` for Voxtream2 models, which means `self._tts_model_type` is `None`, and:

- The `elif self._tts_model_type == "voxtream2"` branch in `_prepare_speech_generation` never executes.
- `_build_voxtream2_prompt` is never called.
- `_validate_voxtream2_request` is never called (it is also not wired into `_validate_tts_request` -- see below).
- The `_load_supported_speakers` early return for voxtream2 is dead.

**Fix:** Add a detection branch in `_detect_tts_model_type`:
```python
if model_stage in _VOXTREAM2_TTS_MODEL_STAGES:
    return "voxtream2"
```

---

### Critical: `_validate_voxtream2_request` is defined but never called

The new `_validate_voxtream2_request` method is not wired into `_validate_tts_request`. This means:
1. In the single-request path (`_prepare_speech_generation`), voxtream2 validation is skipped entirely.
2. In the batch path (`_run_item` calls `_validate_tts_request`), voxtream2 requests fall through to `_validate_qwen_tts_request`, which will reject them with misleading errors.

**Fix:** Add to `_validate_tts_request`:
```python
if self._tts_model_type == "voxtream2":
    return self._validate_voxtream2_request(request)
```

---

### Bug: `generate_stream` type annotation vs. unpacking mismatch

In `voxtream2.py`, `generate_stream` is annotated as returning `Generator[np.ndarray | torch.Tensor, None, None]` (single values). But in `forward()` (line ~1537 of the diff), it is unpacked as a tuple:

```python
for audio_frame, _ in self.generate_stream(...):
```

If the upstream `SpeechGenerator.generate_stream` yields tuples `(audio_chunk, codec_tokens)`, then the type annotation on the wrapper is wrong. If it yields single arrays, the unpacking will crash at runtime. Either way, one of these is wrong.

---

### Regression: VoxCPM2 handling removed

The diff removes three VoxCPM2 code paths:

1. **`_detect_tts_model_type`** (diff line ~1035): Removes `if model_stage in _VOXCPM2_TTS_MODEL_STAGES: return "voxcpm2"`. This specific removal appears harmless because the combined `_VOXCPM_TTS_MODEL_STAGES | _VOXCPM2_TTS_MODEL_STAGES` block above it already handles voxcpm2 detection. However, this removal is still confusing in a PR that is supposed to only add Voxtream2 support.

2. **`_validate_tts_request`** (diff line ~1053): Removes `if self._tts_model_type == "voxcpm2": return None`. This is a real regression for the batch path -- VoxCPM2 batch requests will now fall through to `_validate_qwen_tts_request` and be rejected.

3. **`_prepare_speech_generation`** (diff line ~1197): Removes `elif self._tts_model_type == "voxcpm2": model_type = "voxcpm2"`. This causes VoxCPM2 to log as model_type `"unknown"` instead of `"voxcpm2"`. Minor but unnecessary regression.

**Fix:** Revert these three VoxCPM2 removals. They are unrelated to Voxtream2 support.

---

### Module-level global side effects in `voxtream2.py`

```python
torch.set_float32_matmul_precision("medium")
torch._inductor.config.fx_graph_cache = True
```

These are set at module import time and affect all models in the process, not just Voxtream2. If another model relies on `"high"` matmul precision, this silently degrades its accuracy.

**Fix:** Move these into `__init__` of `Voxtream2ForConditionalGeneration` or gate them behind a Voxtream2-specific env var/config check.

---

### Minor: `_estimate_prompt_len` is duplicated 3 times

The same `max(64, len(text) // 2 + 64)` formula appears in:
- `examples/offline_inference/voxtream2/end2end.py`
- `tests/e2e/offline_inference/test_voxtream2.py`
- `vllm_omni/entrypoints/openai/serving_speech.py` (`_build_voxtream2_prompt`)

Consider extracting this to a shared utility in the voxtream2 model package to avoid drift.

---

### Minor: `import_voxtream_attrs` in `voxtream2_import_utils.py` is unused

The function `import_voxtream_attrs` is defined but never called anywhere in this PR. If it is intended for future use, consider removing it and adding it when needed.

---

### Nit: `_flatten_audio` is also duplicated between `end2end.py` and the test file

Both the example and the test have nearly identical `_flatten_audio` implementations. A shared helper would reduce maintenance burden.

---

### Summary

The model wrapper, config registration, import utilities, stage YAML, tests, and documentation are well-structured. However, the online serving integration has critical wiring bugs that would make Voxtream2 non-functional through the `/v1/audio/speech` endpoint. The unintended VoxCPM2 regressions should also be reverted. Please fix the issues above and verify the online serving path end-to-end before merging.

**Comment by ccczzc** [comment]
@lishunyang12 Thanks for the valuable review. I’ve addressed all the comments in the latest commit.

**Comment by BeatSeat** [comment]
Since MiniCPM-o4.5 is supported, are you interested in refactor to the full-duplex later?
